leopard.nim

```diff
@@ -0,0 +1,329 @@
+  proc c_free(p: pointer) {.header: "<stdlib.h>", importc: "free".}
+
+  proc freeAligned*(p: pointer) {.inline.} =
+    if not p.isNil:
+      c_free(p)
+
+proc allocAligned*(size: int, alignment: static Natural): pointer {.inline.} =
+  ## aligned_alloc requires allocating in multiple of the alignment.
+  static:
+    assert alignment.isPowerOfTwo()
+
+  let
+    requiredMem = size.roundNextMultipleOf(alignment)
+
+  when defined(windows):
+    aligned_alloc_windows(csize_t requiredMem, csize_t alignment)
+  else:
+    aligned_alloc(csize_t alignment, csize_t requiredMem)
+
+proc leoInit*() =
+  if wrapper.leoInit() != 0:
+    raise (ref LeopardDefect)(msg: "Leopard-RS failed to initialize")
+
+proc encode*(code: ReedSolomonCode, data: Data):
+    Result[ParityData, LeopardError] =
+  if not code.isValid:
+    return err LeopardError(code: LeopardBadCode, msg: LeopardBadCodeMsg)
+
+  var
+    data = data
+
+  let
+    symbolBytes = data[0].len
+
+  if data.len < code.data.int:
+    return err LeopardError(code: LeopardNotEnoughData,
+      msg: LeopardNotEnoughDataMsg)
+
+  elif data.len > code.data.int:
+    return err LeopardError(code: LeopardTooMuchData,
+      msg: $leoResultString(wrapper.LeopardTooMuchData))
+
+  if symbolBytes < MinBufferSize.int or symbolBytes mod MinBufferSize.int != 0:
+    return err LeopardError(code: LeopardInvalidSize,
+      msg: $leoResultString(wrapper.LeopardInvalidSize))
+
+  var
+    enData = newSeq[pointer](code.data)
+
+  for i in 0..<code.data:
+    if data[i].len != symbolBytes:
+      for i in 0..<code.data: freeAligned enData[i]
+      return err LeopardError(code: LeopardInconsistentSize,
+        msg: LeopardInconsistentSizeMsg)
+
+    enData[i] = allocAligned(symbolBytes, LEO_ALIGN_BYTES)
+    moveMem(enData[i], addr data[i][0], symbolBytes)
+
+  let
+    workCount = leoEncodeWorkCount(code.data.cuint, code.parity.cuint)
+
+  if workCount == 0:
+    for i in 0..<code.data: freeAligned enData[i]
+    return err LeopardError(code: LeopardInvalidInput,
+      msg: $leoResultString(wrapper.LeopardInvalidInput))
+
+  var
+    workData = newSeq[pointer](workCount)
+
+  for i in 0..<workCount:
+    workData[i] = allocAligned(symbolBytes, LEO_ALIGN_BYTES)
+
+  let
+    encodeRes = leoEncode(
+      symbolBytes.uint64,
+      code.data.cuint,
+      code.parity.cuint,
+      workCount,
+      addr enData[0],
+      addr workData[0]
+    )
+
+  if encodeRes != wrapper.LeopardSuccess:
+    for i in 0..<code.data: freeAligned enData[i]
+    for i in 0..<workCount: freeAligned workData[i]
+    return err LeopardError(code: cast[LeopardResult](encodeRes),
+      msg: $leoResultString(encodeRes))
+
+  var
+    parityData: ParityData
+
+  newSeq(parityData, code.parity)
+  for i in 0..<code.parity:
+    newSeq(parityData[i], symbolBytes)
+    moveMem(addr parityData[i][0], workData[i], symbolBytes)
+
+  for i in 0..<code.data: freeAligned enData[i]
+  for i in 0..<workCount: freeAligned workData[i]
```

Comment:
Why free again?

Reply:
All the freeing above was done only in the case of early returns (error conditions).

The freeing on L212-L213 is done just prior to successful return.